kernel/Utilities/Eigenpairs.cpp

```diff
@@ -0,0 +1,131 @@
+ this list of conditions and the following disclaimer in the documentation
+ and/or other materials provided with the distribution.
+
+ 3. Neither the name of the copyright holder nor the names of its contributors
+ may be used to endorse or promote products derived from this software without
+ specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifdef HPGEM_USE_SLEPC
+
+#include <numeric>
+#include "Eigenpairs.h"
+
+namespace hpgem {
+namespace Utilities {
+Eigenpairs::Eigenpairs()
+    : numberOfEigenvectors_(0),
+      eigenvectors_(nullptr),
+      eigenvalues_(0),
+      ordering_(0){};
+
+Eigenpairs::Eigenpairs(Eigenpairs &&other) noexcept
+    : numberOfEigenvectors_(0),
+      eigenvectors_(nullptr),
+      eigenvalues_(0),
+      ordering_(0) {
+    // Essential, as the other vector now does not have any eigenvalues
+    std::swap(numberOfEigenvectors_, other.numberOfEigenvectors_);
+    std::swap(eigenvectors_, other.eigenvectors_);
+    std::swap(eigenvalues_, other.eigenvalues_);
+    std::swap(ordering_, other.ordering_);
+}
+
+Eigenpairs::~Eigenpairs() {
+    if (numberOfEigenvectors_ > 0) {
+        PetscErrorCode err =
+            VecDestroyVecs(numberOfEigenvectors_, &eigenvectors_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+    }
+}
+
+Eigenpairs &Eigenpairs::operator=(Eigenpairs &&other) noexcept {
+    std::swap(numberOfEigenvectors_, other.numberOfEigenvectors_);
+    std::swap(eigenvectors_, other.eigenvectors_);
+    std::swap(eigenvalues_, other.eigenvalues_);
+    std::swap(ordering_, other.ordering_);
+    return *this;
+}
+
+void Eigenpairs::reserve(std::size_t newSize, Vec sample) {
+    if (newSize <= numberOfEigenvectors_) {
+        return;
+    }
+    PetscErrorCode err = VecDestroyVecs(numberOfEigenvectors_, &eigenvectors_);
+    CHKERRABORT(PETSC_COMM_WORLD, err);
+    err = VecDuplicateVecs(sample, newSize, &eigenvectors_);
+    CHKERRABORT(PETSC_COMM_WORLD, err);
+    numberOfEigenvectors_ = newSize;
+}
+
+/**
+ * \brief Load eigenpairs from a solver
+ *
+ * @param eps The solver to load the eigenpairs from
+ * @param sample A sample vector to duplicate for storing eigenvectors
```

Comment:
I do not understand the idea behind sample, it is a vector with the right shape, which is not used?

Reply:
It is a vector which has the correct type (e.g. not a block vector) and layout (the number of entries per processor). The content of the vector is not used, but it is used as a cloning template. By cloning it we get the actual storage for the eigenvectors.